python/ray/serve/_private/deployment_state.py

```diff
@@ -1465,88 +1459,59 @@ def _set_target_state(
-        if existing_info is not None:
+
+        # Exit early if the deployment info hasn't changed. Ensures this method
+        # is idempotent.
+        curr_deployment_info = self._target_state.info
+        if curr_deployment_info is not None:
             # Redeploying should not reset the deployment's start time.
             if not self._target_state.deleting:
-                deployment_info.start_time_ms = existing_info.start_time_ms
+                deployment_info.start_time_ms = curr_deployment_info.start_time_ms
 
             if (
                 not self._target_state.deleting
-                and existing_info.deployment_config == deployment_info.deployment_config
-                and existing_info.replica_config.ray_actor_options
+                and curr_deployment_info.deployment_config
+                == deployment_info.deployment_config
+                and curr_deployment_info.replica_config.ray_actor_options
                 == deployment_info.replica_config.ray_actor_options
                 and deployment_info.version is not None
-                and existing_info.version == deployment_info.version
+                and curr_deployment_info.version == deployment_info.version
+                and curr_deployment_info.target_capacity
+                == deployment_info.target_capacity
+                and curr_deployment_info.target_capacity_direction
+                == deployment_info.target_capacity_direction
             ):
                 return False
 
-        # If autoscaling config is not none, decide initial num replicas
-        autoscaling_config = deployment_info.deployment_config.autoscaling_config
-        if autoscaling_config is not None:
-            adjust_capacity = False
-            if autoscaling_config.initial_replicas is not None and (
-                target_capacity_scale_direction is None
-                or target_capacity_scale_direction == TargetCapacityScaleDirection.UP
-            ):
-                autoscaled_num_replicas = autoscaling_config.initial_replicas
+        # Decide new target num_replicas.
+        autoscaling_policy = deployment_info.autoscaling_policy
+        if autoscaling_policy is not None:
+            if curr_deployment_info is None:
+                target_num_replicas = autoscaling_policy.apply_initial_bounds(-1)
             else:
-                if existing_info is not None:
-                    autoscaled_num_replicas = self._target_state.num_replicas
-                else:
-                    autoscaled_num_replicas = autoscaling_config.min_replicas
-            autoscaled_num_replicas = self.get_capacity_adjusted_num_replicas(
-                autoscaled_num_replicas, target_capacity
-            )
-            deployment_info.set_autoscaled_num_replicas(autoscaled_num_replicas)
+                target_num_replicas = autoscaling_policy.apply_initial_bounds(
+                    self._target_state.target_num_replicas
```

Comment:
I'm struggling to understand how this interacts with `get_decision_num_replicas` -- why are both needed and the calls made in different places?

Reply:
I've refactored this logic a bit. See [this reply](https://github.com/ray-project/ray/pull/41317#discussion_r1409841115) for links.